added documents marked as to be removed clean script (file_garbage_remover)

idelcano opened the pull request:
This script removes documents that should have been deleted by DHIS2 but weren't.

to use
python3 file_garbage_remover.py -i instancename

Introduced file_garbage_remover_tomcat.py:
A robust script designed specifically for production (Tomcat) environments. This script safely handles orphaned document resources by moving them to a temporary directory, creating an audit trail in the database, and then cleaning up original entries. It provides test (--test) and force (--force) modes for secure operations.

Updated Documentation:
Clearly documented both scripts (file_garbage_remover_tomcat.py and the previously existing Docker test script) to distinguish their usage, operation, and precautions.

This improvement allows safer file handling and audit tracking in production, complementing the simpler Docker-based script intended for testing purposes.

I have create a wrapper in dev to test it in dry run mode:

`/home/tomcatuser/bin/file_garbage_remover/run_file_garbage_remover.sh test `

idelcano committed:
added documents marked as to be removed clean script

cgbautista commented on DHIS2/post_clone_scripts/file_garbage_remover.py:
```diff
@@ -0,0 +1,110 @@
+
+
+def find_container_id(name_pattern):
+    """Find the container ID based on a name substring."""
+    result = run(["docker", "ps", "--format", "{{.ID}}\t{{.Names}}"], capture=True)
+    for line in result.stdout.strip().splitlines():
+        cid, name = line.split("\t")
+        if name_pattern in name:
+            return cid
+    return None
```
To avoid having to manually filter, just ask the command to give us the filtered list
```suggestion
    """Find the container ID based on a name substring. Retrieves first match only"""
    result = run(["docker", "ps", "--format", "{{.ID}}", "-f", f"name={name_pattern}"], capture=True)
    lines = result.stdout.strip().splitlines()
    return lines[0] if lines else None
```

cgbautista commented on DHIS2/post_clone_scripts/file_garbage_remover.py:
```diff
@@ -0,0 +1,110 @@
+
+    instance_name = args.instance
+    container_slug = slugify(instance_name)
+    db_container_match = container_slug + "-db-1"
+    core_container_match = container_slug + "-core-1"
```
Deleted db_containter_match as it is not needed
Moved the replace to where the match string is constructed.
```suggestion
    core_container_match = container_slug.replace("dhis2-data-", "") + "-core-1"
```

cgbautista commented on DHIS2/post_clone_scripts/file_garbage_remover.py:
```diff
@@ -0,0 +1,110 @@
+                f.write(f"{line}\n")
+
+    print("Identifying containers...")
+    db_container = find_container_id(db_container_match.replace("dhis2-data-", ""))
+    core_container = find_container_id(core_container_match.replace("dhis2-data-",""))
+    print(db_container)
+    print(core_container)
+    if not db_container or not core_container:
+        print("Could not find DB or Core container.")
+        print(f"Looked for: {db_container_match}, {core_container_match}")
+        sys.exit(1)
+
+    print(f"DB container: {db_container}")
+    print(f"Core container: {core_container}")
```
Removed db_container as it is not used. 
Search for the match string directly
```suggestion
    core_container = find_container_id(core_container_match)
    print(core_container)
    if not core_container:
        print("Could not find Core container.")
        print(f"Looked for: {core_container_match}")
        sys.exit(1)

    print(f"Core container: {core_container}")
```

cgbautista commented on DHIS2/post_clone_scripts/file_garbage_remover.py:
```diff
@@ -0,0 +1,110 @@
+    print("Copying file list to Core container...")
+    run(["docker", "cp", LOCAL_LIST_FILE, f"{core_container}:{REMOTE_LIST_FILE}"])
+
+    print(f"Deleting orphaned files in Core container... {core_container} {db_container_match}")
```
```suggestion
    print(f"Deleting orphaned files in Core container... {core_container}")
```

cgbautista reviewed (approved):
Looks good from my side. Suggested some changes, but would work as is.

idelcano committed:
rename folder

idelcano committed:
remove unnecesary code and simplify script

idelcano committed:
rename file

idelcano committed:
Added tomcat version

idelcano committed:
remove duplicate old file

idelcano committed:
Added readme

idelcano commented:
Hi @cgbautista 
Please review the first comment and the updated README.md. If you have any questions, feel free to ask.
This script has already been executed in the preprod-indiv and dev environments as of this comment (we could think that the functional testing is done).

cgbautista commented on DHIS2/file_garbage_remover/Readme.md:
````diff
@@ -0,0 +1,72 @@
+
+Test Mode (dry-run):
+
+```./file_garbage_remover_tomcat.py --test --config /path/to/config.json```
````
Test mode also requires the db password to connect and retrieve the list of fileResources, so the code example should contain the `export` line as the _--force_ example

cgbautista commented on DHIS2/file_garbage_remover/file_garbage_remover_tomcat.py:
```diff
@@ -0,0 +1,181 @@
+def ensure_audit_table_exists(cursor, dry_run=False):
+    if dry_run:
+        log("[DRY RUN] Would execute:")
+        log(" CREATE TABLE IF NOT EXISTS fileresourcesaudit AS TABLE fileresource WITH NO DATA;")
```
I would include this SQL query (and the same query in line 109) the same way as the other 3 queries in lines 12-30

cgbautista commented on DHIS2/file_garbage_remover/file_garbage_remover_tomcat.py:
```diff
@@ -0,0 +1,181 @@
+
+    db_password = os.environ.get("DB_PASSWORD_FILE_G")
+    if not db_password:
+        log("❌ Missing required environment variable: DB_PASSWORD")
```
Error message informs that the user did not inform an environmental variable that is not the same that the code checks.
If the password is expected to be as plain text in the env., it should not be called _FILE_

idelcano replied:
Yes, the name was really bad
I used it to avoid introducing another DB_PASSWORD environment variable (definitely improvable) and was thinking in FILE_GARBAGE.py, but it doesn't make sense. I've renamed it to DB_PASSWORD_FG.

idelcano committed:
fix readme

idelcano committed:
Refactoring to remove magic values and fix log comment

idelcano committed:
change db variable name

idelcano commented:
Thanks @cgbautista I did a new commits to fix this minus issues

cgbautista committed:
Edited error reported to the user to reflect the correct name of the environmental variable

cgbautista reviewed (approved):
Tested in local enviroment both on docker and tomcat instances.

idelcano committed:
Update script to clean data value file resources excluding actives in event, datavalue, trackedentityattribute values.

idelcano committed:
update to add notifications and refactor

idelcano committed:
improve script fixing error, adding notify, improve workflow and fix docker remover

idelcano committed:
refactor sql

idelcano committed:
read proxy from config file

idelcano committed:
added store files in csv as already notified parameter and truncate messages

idelcano committed:
refactor duplicate code

idelcano committed:
refactor

idelcano committed:
simplify parameters and workflod, improve unique row detection(id+uid), and add save-all-as-notified

idelcano committed:
import missing dependency

idelcano committed:
avoid identify as orphan files created/updated in the last 24

idelcano committed:
recovery instructions

idelcano committed:
Add ignore-logger-files filter for JSON timestamp logs and better get/set id

idelcano committed:
added new parameter to ignore log files (created by user-extended-app) to script

idelcano commented:
Hi @cgbautista , I added several commits to the file garbage remover with big changes from your review. 
Could you review it again? 
It's already working in our instances.

Changes summary (grouped):

Orphan detection: added 24h safety window; improved SQL for data value file resources; refactor of SQL queries.
Notifications: added notify flow, webhook support, proxy from config, notify-only mode, truncation, and “save all as notified.”
CSV handling: dedup based on id+uid, merge notified flags, and workflow improvements to avoid re-notifying.
Refactors: removed duplicate code, simplified parameters, and general cleanup.
New: ignore logger JSON timestamp files via --ignore-logger-files.
Fixed d2-docker. It work with tomcat/d2-docker now.
